// @wordpress/block-editor/build/components/image-editor/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ImageEditor;

var _element = require("@wordpress/element");

var _components = require("@wordpress/components");

var _blockControls = _interopRequireDefault(require("../block-controls"));

var _context = _interopRequireDefault(require("./context"));

var _cropper = _interopRequireDefault(require("./cropper"));

var _zoomDropdown = _interopRequireDefault(require("./zoom-dropdown"));

var _aspectRatioDropdown = _interopRequireDefault(require("./aspect-ratio-dropdown"));

var _rotationButton = _interopRequireDefault(require("./rotation-button"));

var _formControls = _interopRequireDefault(require("./form-controls"));

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
function ImageEditor(_ref) {
  let {
    id,
    url,
    width,
    height,
    clientWidth,
    naturalHeight,
    naturalWidth,
    onSaveImage,
    onFinishEditing,
    borderProps
  } = _ref;
  return (0, _element.createElement)(_context.default, {
    id: id,
    url: url,
    naturalWidth: naturalWidth,
    naturalHeight: naturalHeight,
    onSaveImage: onSaveImage,
    onFinishEditing: onFinishEditing
  }, (0, _element.createElement)(_cropper.default, {
    borderProps: borderProps,
    url: url,
    width: width,
    height: height,
    clientWidth: clientWidth,
    naturalHeight: naturalHeight,
    naturalWidth: naturalWidth
  }), (0, _element.createElement)(_blockControls.default, null, (0, _element.createElement)(_components.ToolbarGroup, null, (0, _element.createElement)(_zoomDropdown.default, null), (0, _element.createElement)(_components.ToolbarItem, null, toggleProps => (0, _element.createElement)(_aspectRatioDropdown.default, {
    toggleProps: toggleProps
  })), (0, _element.createElement)(_rotationButton.default, null)), (0, _element.createElement)(_components.ToolbarGroup, null, (0, _element.createElement)(_formControls.default, null))));
}
