// @wordpress/block-editor/build/components/image-editor/zoom-dropdown.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ZoomDropdown;

var _element = require("@wordpress/element");

var _components = require("@wordpress/components");

var _i18n = require("@wordpress/i18n");

var _icons = require("@wordpress/icons");

var _constants = require("./constants");

var _context = require("./context");

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
function ZoomDropdown() {
  const {
    isInProgress,
    zoom,
    setZoom
  } = (0, _context.useImageEditingContext)();
  return (0, _element.createElement)(_components.Dropdown, {
    contentClassName: "wp-block-image__zoom",
    popoverProps: _constants.POPOVER_PROPS,
    renderToggle: _ref => {
      let {
        isOpen,
        onToggle
      } = _ref;
      return (0, _element.createElement)(_components.ToolbarButton, {
        icon: _icons.search,
        label: (0, _i18n.__)('Zoom'),
        onClick: onToggle,
        "aria-expanded": isOpen,
        disabled: isInProgress
      });
    },
    renderContent: () => (0, _element.createElement)(_components.RangeControl, {
      __nextHasNoMarginBottom: true,
      label: (0, _i18n.__)('Zoom'),
      min: _constants.MIN_ZOOM,
      max: _constants.MAX_ZOOM,
      value: Math.round(zoom),
      onChange: setZoom
    })
  });
}
