// @wordpress/block-editor/build/components/link-control/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.URL_TYPE = exports.TEL_TYPE = exports.MAILTO_TYPE = exports.LINK_ENTRY_TYPES = exports.INTERNAL_TYPE = exports.DEFAULT_LINK_SETTINGS = exports.CREATE_TYPE = void 0;

var _i18n = require("@wordpress/i18n");

/**
 * WordPress dependencies
 */
// Used as a unique identifier for the "Create" option within search results.
// Used to help distinguish the "Create" suggestion within the search results in
// order to handle it as a unique case.
const CREATE_TYPE = '__CREATE__';
exports.CREATE_TYPE = CREATE_TYPE;
const TEL_TYPE = 'tel';
exports.TEL_TYPE = TEL_TYPE;
const URL_TYPE = 'URL';
exports.URL_TYPE = URL_TYPE;
const MAILTO_TYPE = 'mailto';
exports.MAILTO_TYPE = MAILTO_TYPE;
const INTERNAL_TYPE = 'internal';
exports.INTERNAL_TYPE = INTERNAL_TYPE;
const LINK_ENTRY_TYPES = [URL_TYPE, MAILTO_TYPE, TEL_TYPE, INTERNAL_TYPE];
exports.LINK_ENTRY_TYPES = LINK_ENTRY_TYPES;
const DEFAULT_LINK_SETTINGS = [{
  id: 'opensInNewTab',
  title: (0, _i18n.__)('Open in new tab')
}];
exports.DEFAULT_LINK_SETTINGS = DEFAULT_LINK_SETTINGS;
