// @wordpress/block-editor/build/components/global-styles/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "DimensionsPanel", {
  enumerable: true,
  get: function () {
    return _dimensionsPanel.default;
  }
});
Object.defineProperty(exports, "GlobalStylesContext", {
  enumerable: true,
  get: function () {
    return _context.GlobalStylesContext;
  }
});
Object.defineProperty(exports, "TypographyPanel", {
  enumerable: true,
  get: function () {
    return _typographyPanel.default;
  }
});
Object.defineProperty(exports, "useGlobalSetting", {
  enumerable: true,
  get: function () {
    return _hooks.useGlobalSetting;
  }
});
Object.defineProperty(exports, "useGlobalStyle", {
  enumerable: true,
  get: function () {
    return _hooks.useGlobalStyle;
  }
});
Object.defineProperty(exports, "useGlobalStylesOutput", {
  enumerable: true,
  get: function () {
    return _useGlobalStylesOutput.useGlobalStylesOutput;
  }
});
Object.defineProperty(exports, "useGlobalStylesReset", {
  enumerable: true,
  get: function () {
    return _hooks.useGlobalStylesReset;
  }
});
Object.defineProperty(exports, "useHasDimensionsPanel", {
  enumerable: true,
  get: function () {
    return _dimensionsPanel.useHasDimensionsPanel;
  }
});
Object.defineProperty(exports, "useHasTypographyPanel", {
  enumerable: true,
  get: function () {
    return _typographyPanel.useHasTypographyPanel;
  }
});
Object.defineProperty(exports, "useSettingsForBlockElement", {
  enumerable: true,
  get: function () {
    return _hooks.useSettingsForBlockElement;
  }
});

var _hooks = require("./hooks");

var _useGlobalStylesOutput = require("./use-global-styles-output");

var _context = require("./context");

var _typographyPanel = _interopRequireWildcard(require("./typography-panel"));

var _dimensionsPanel = _interopRequireWildcard(require("./dimensions-panel"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
