// @wordpress/block-library/build/comments-pagination-previous/edit.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = CommentsPaginationPreviousEdit;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _i18n = require("@wordpress/i18n");

var _blockEditor = require("@wordpress/block-editor");

/**
 * WordPress dependencies
 */
const arrowMap = {
  none: '',
  arrow: '←',
  chevron: '«'
};

function CommentsPaginationPreviousEdit(_ref) {
  let {
    attributes: {
      label
    },
    setAttributes,
    context: {
      'comments/paginationArrow': paginationArrow
    }
  } = _ref;
  const displayArrow = arrowMap[paginationArrow];
  return (0, _element.createElement)("a", (0, _extends2.default)({
    href: "#comments-pagination-previous-pseudo-link",
    onClick: event => event.preventDefault()
  }, (0, _blockEditor.useBlockProps)()), displayArrow && (0, _element.createElement)("span", {
    className: `wp-block-comments-pagination-previous-arrow is-arrow-${paginationArrow}`
  }, displayArrow), (0, _element.createElement)(_blockEditor.PlainText, {
    __experimentalVersion: 2,
    tagName: "span",
    "aria-label": (0, _i18n.__)('Older comments page link'),
    placeholder: (0, _i18n.__)('Older Comments'),
    value: label,
    onChange: newLabel => setAttributes({
      label: newLabel
    })
  }));
}
