// @wordpress/block-editor/build/components/inserter/block-patterns-explorer/explorer.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _components = require("@wordpress/components");

var _i18n = require("@wordpress/i18n");

var _sidebar = _interopRequireDefault(require("./sidebar"));

var _patternsList = _interopRequireDefault(require("./patterns-list"));

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
function PatternsExplorer(_ref) {
  let {
    initialCategory,
    patternCategories
  } = _ref;
  const [filterValue, setFilterValue] = (0, _element.useState)('');
  const [selectedCategory, setSelectedCategory] = (0, _element.useState)(initialCategory === null || initialCategory === void 0 ? void 0 : initialCategory.name);
  return (0, _element.createElement)("div", {
    className: "block-editor-block-patterns-explorer"
  }, (0, _element.createElement)(_sidebar.default, {
    selectedCategory: selectedCategory,
    patternCategories: patternCategories,
    onClickCategory: setSelectedCategory,
    filterValue: filterValue,
    setFilterValue: setFilterValue
  }), (0, _element.createElement)(_patternsList.default, {
    filterValue: filterValue,
    selectedCategory: selectedCategory,
    patternCategories: patternCategories
  }));
}

function PatternsExplorerModal(_ref2) {
  let {
    onModalClose,
    ...restProps
  } = _ref2;
  return (0, _element.createElement)(_components.Modal, {
    title: (0, _i18n.__)('Patterns'),
    onRequestClose: onModalClose,
    isFullScreen: true
  }, (0, _element.createElement)(PatternsExplorer, restProps));
}

var _default = PatternsExplorerModal;
exports.default = _default;
