// @wordpress/block-editor/build/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  __experimentalGetBorderClassesAndStyles: true,
  __experimentalUseBorderProps: true,
  __experimentalGetColorClassesAndStyles: true,
  getTypographyClassesAndStyles: true,
  __experimentalUseColorProps: true,
  __experimentalUseCustomSides: true,
  __experimentalGetSpacingClassesAndStyles: true,
  __experimentalGetGapCSSValue: true,
  useCachedTruthy: true,
  __experimentaluseLayoutClasses: true,
  __experimentaluseLayoutStyles: true,
  storeConfig: true,
  store: true,
  SETTINGS_DEFAULTS: true,
  privateApis: true
};
Object.defineProperty(exports, "SETTINGS_DEFAULTS", {
  enumerable: true,
  get: function () {
    return _defaults.SETTINGS_DEFAULTS;
  }
});
Object.defineProperty(exports, "__experimentalGetBorderClassesAndStyles", {
  enumerable: true,
  get: function () {
    return _hooks.getBorderClassesAndStyles;
  }
});
Object.defineProperty(exports, "__experimentalGetColorClassesAndStyles", {
  enumerable: true,
  get: function () {
    return _hooks.getColorClassesAndStyles;
  }
});
Object.defineProperty(exports, "__experimentalGetGapCSSValue", {
  enumerable: true,
  get: function () {
    return _hooks.getGapCSSValue;
  }
});
Object.defineProperty(exports, "__experimentalGetSpacingClassesAndStyles", {
  enumerable: true,
  get: function () {
    return _hooks.getSpacingClassesAndStyles;
  }
});
Object.defineProperty(exports, "__experimentalUseBorderProps", {
  enumerable: true,
  get: function () {
    return _hooks.useBorderProps;
  }
});
Object.defineProperty(exports, "__experimentalUseColorProps", {
  enumerable: true,
  get: function () {
    return _hooks.useColorProps;
  }
});
Object.defineProperty(exports, "__experimentalUseCustomSides", {
  enumerable: true,
  get: function () {
    return _hooks.useCustomSides;
  }
});
Object.defineProperty(exports, "__experimentaluseLayoutClasses", {
  enumerable: true,
  get: function () {
    return _hooks.useLayoutClasses;
  }
});
Object.defineProperty(exports, "__experimentaluseLayoutStyles", {
  enumerable: true,
  get: function () {
    return _hooks.useLayoutStyles;
  }
});
Object.defineProperty(exports, "getTypographyClassesAndStyles", {
  enumerable: true,
  get: function () {
    return _hooks.getTypographyClassesAndStyles;
  }
});
Object.defineProperty(exports, "privateApis", {
  enumerable: true,
  get: function () {
    return _privateApis.privateApis;
  }
});
Object.defineProperty(exports, "store", {
  enumerable: true,
  get: function () {
    return _store.store;
  }
});
Object.defineProperty(exports, "storeConfig", {
  enumerable: true,
  get: function () {
    return _store.storeConfig;
  }
});
Object.defineProperty(exports, "useCachedTruthy", {
  enumerable: true,
  get: function () {
    return _hooks.useCachedTruthy;
  }
});

var _hooks = require("./hooks");

var _components = require("./components");

Object.keys(_components).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _components[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _components[key];
    }
  });
});

var _elements = require("./elements");

Object.keys(_elements).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _elements[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _elements[key];
    }
  });
});

var _utils = require("./utils");

Object.keys(_utils).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _utils[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _utils[key];
    }
  });
});

var _store = require("./store");

var _defaults = require("./store/defaults");

var _privateApis = require("./private-apis");
