// @wordpress/block-library/build/gallery/gallery.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Gallery = void 0;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classnames = _interopRequireDefault(require("classnames"));

var _blockEditor = require("@wordpress/block-editor");

var _i18n = require("@wordpress/i18n");

var _blocks = require("@wordpress/blocks");

var _primitives = require("@wordpress/primitives");

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */
const Gallery = (props, captionRef) => {
  const {
    attributes,
    isSelected,
    setAttributes,
    mediaPlaceholder,
    insertBlocksAfter,
    blockProps,
    __unstableLayoutClassNames: layoutClassNames,
    showCaption
  } = props;
  const {
    align,
    columns,
    caption,
    imageCrop
  } = attributes;
  return (0, _element.createElement)("figure", (0, _extends2.default)({}, blockProps, {
    className: (0, _classnames.default)(blockProps.className, layoutClassNames, 'blocks-gallery-grid', {
      [`align${align}`]: align,
      [`columns-${columns}`]: columns !== undefined,
      [`columns-default`]: columns === undefined,
      'is-cropped': imageCrop
    })
  }), blockProps.children, isSelected && !blockProps.children && (0, _element.createElement)(_primitives.View, {
    className: "blocks-gallery-media-placeholder-wrapper"
  }, mediaPlaceholder), showCaption && (!_blockEditor.RichText.isEmpty(caption) || isSelected) && (0, _element.createElement)(_blockEditor.RichText, {
    identifier: "caption",
    "aria-label": (0, _i18n.__)('Gallery caption text'),
    placeholder: (0, _i18n.__)('Write gallery caption…'),
    value: caption,
    className: (0, _classnames.default)('blocks-gallery-caption', (0, _blockEditor.__experimentalGetElementClassName)('caption')),
    ref: captionRef,
    tagName: "figcaption",
    onChange: value => setAttributes({
      caption: value
    }),
    inlineToolbar: true,
    __unstableOnSplitAtEnd: () => insertBlocksAfter((0, _blocks.createBlock)((0, _blocks.getDefaultBlockName)()))
  }));
};

exports.Gallery = Gallery;

var _default = (0, _element.forwardRef)(Gallery);

exports.default = _default;
