// @wordpress/block-editor/build/components/block-list/block-list-compact.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _reactNative = require("react-native");

var _blockEditor = require("@wordpress/block-editor");

var _data = require("@wordpress/data");

var _style = _interopRequireDefault(require("./style.scss"));

var _block = _interopRequireDefault(require("./block"));

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

/**
 * NOTE: This is a component currently used by the List block (V2)
 * It only passes the needed props for this block, if other blocks will use it
 * make sure you pass other props that might be required coming from:
 * components/inner-blocks/index.native.js
 */
function BlockListCompact(props) {
  const {
    marginHorizontal = _style.default.defaultBlock.marginLeft,
    marginVertical = _style.default.defaultBlock.marginTop,
    rootClientId
  } = props;
  const {
    blockClientIds
  } = (0, _data.useSelect)(select => {
    const {
      getBlockOrder
    } = select(_blockEditor.store);
    const blockOrder = getBlockOrder(rootClientId);
    return {
      blockClientIds: blockOrder
    };
  }, [rootClientId]);
  const containerStyle = {
    marginVertical: -marginVertical,
    marginHorizontal: -marginHorizontal
  };
  return (0, _element.createElement)(_reactNative.View, {
    style: containerStyle,
    testID: "block-list-wrapper"
  }, blockClientIds.map(currentClientId => (0, _element.createElement)(_block.default, {
    clientId: currentClientId,
    rootClientId: rootClientId,
    key: currentClientId,
    marginHorizontal: marginHorizontal,
    marginVertical: marginVertical
  })));
}

var _default = BlockListCompact;
exports.default = _default;
