// @wordpress/block-editor/build-module/components/list-view/leaf.js
import _extends from "@babel/runtime/helpers/esm/extends";
import { createElement } from "@wordpress/element";

/**
 * External dependencies
 */
import { animated } from '@react-spring/web';
import classnames from 'classnames';
/**
 * WordPress dependencies
 */

import { __experimentalTreeGridRow as TreeGridRow } from '@wordpress/components';
import { useMergeRefs } from '@wordpress/compose';
import { forwardRef } from '@wordpress/element';
/**
 * Internal dependencies
 */

import useMovingAnimation from '../use-moving-animation';
const AnimatedTreeGridRow = animated(TreeGridRow);
const ListViewLeaf = forwardRef((_ref, ref) => {
  let {
    isSelected,
    position,
    level,
    rowCount,
    children,
    className,
    path,
    ...props
  } = _ref;
  const animationRef = useMovingAnimation({
    isSelected,
    adjustScrolling: false,
    enableAnimation: true,
    triggerAnimationOnChange: path
  });
  const mergedRef = useMergeRefs([ref, animationRef]);
  return createElement(AnimatedTreeGridRow, _extends({
    ref: mergedRef,
    className: classnames('block-editor-list-view-leaf', className),
    level: level,
    positionInSet: position,
    setSize: rowCount
  }, props), children);
});
export default ListViewLeaf;
