// @wordpress/block-library/build/comments/edit/placeholder.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = PostCommentsPlaceholder;

var _element = require("@wordpress/element");

var _blockEditor = require("@wordpress/block-editor");

var _i18n = require("@wordpress/i18n");

var _data = require("@wordpress/data");

var _coreData = require("@wordpress/core-data");

var _form = _interopRequireDefault(require("../../post-comments-form/form"));

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
function PostCommentsPlaceholder(_ref) {
  let {
    postType,
    postId
  } = _ref;
  let [postTitle] = (0, _coreData.useEntityProp)('postType', postType, 'title', postId);
  postTitle = postTitle || (0, _i18n.__)('Post Title');
  const {
    avatarURL
  } = (0, _data.useSelect)(select => select(_blockEditor.store).getSettings().__experimentalDiscussionSettings);
  return (0, _element.createElement)("div", {
    className: "wp-block-comments__legacy-placeholder",
    inert: "true"
  }, (0, _element.createElement)("h3", null,
  /* translators: %s: Post title. */
  (0, _i18n.sprintf)((0, _i18n.__)('One response to %s'), postTitle)), (0, _element.createElement)("div", {
    className: "navigation"
  }, (0, _element.createElement)("div", {
    className: "alignleft"
  }, (0, _element.createElement)("a", {
    href: "#top"
  }, "\xAB ", (0, _i18n.__)('Older Comments'))), (0, _element.createElement)("div", {
    className: "alignright"
  }, (0, _element.createElement)("a", {
    href: "#top"
  }, (0, _i18n.__)('Newer Comments'), " \xBB"))), (0, _element.createElement)("ol", {
    className: "commentlist"
  }, (0, _element.createElement)("li", {
    className: "comment even thread-even depth-1"
  }, (0, _element.createElement)("article", {
    className: "comment-body"
  }, (0, _element.createElement)("footer", {
    className: "comment-meta"
  }, (0, _element.createElement)("div", {
    className: "comment-author vcard"
  }, (0, _element.createElement)("img", {
    alt: "Commenter Avatar",
    src: avatarURL,
    className: "avatar avatar-32 photo",
    height: "32",
    width: "32",
    loading: "lazy"
  }), (0, _element.createElement)("b", {
    className: "fn"
  }, (0, _element.createElement)("a", {
    href: "#top",
    className: "url"
  }, (0, _i18n.__)('A WordPress Commenter'))), ' ', (0, _element.createElement)("span", {
    className: "says"
  }, (0, _i18n.__)('says'), ":")), (0, _element.createElement)("div", {
    className: "comment-metadata"
  }, (0, _element.createElement)("a", {
    href: "#top"
  }, (0, _element.createElement)("time", {
    dateTime: "2000-01-01T00:00:00+00:00"
  }, (0, _i18n.__)('January 1, 2000 at 00:00 am'))), ' ', (0, _element.createElement)("span", {
    className: "edit-link"
  }, (0, _element.createElement)("a", {
    className: "comment-edit-link",
    href: "#top"
  }, (0, _i18n.__)('Edit'))))), (0, _element.createElement)("div", {
    className: "comment-content"
  }, (0, _element.createElement)("p", null, (0, _i18n.__)('Hi, this is a comment.'), (0, _element.createElement)("br", null), (0, _i18n.__)('To get started with moderating, editing, and deleting comments, please visit the Comments screen in the dashboard.'), (0, _element.createElement)("br", null), (0, _i18n.__)('Commenter avatars come from'), ' ', (0, _element.createElement)("a", {
    href: "https://gravatar.com/"
  }, "Gravatar"), ".")), (0, _element.createElement)("div", {
    className: "reply"
  }, (0, _element.createElement)("a", {
    className: "comment-reply-link",
    href: "#top",
    "aria-label": "Reply to A WordPress Commenter"
  }, (0, _i18n.__)('Reply')))))), (0, _element.createElement)("div", {
    className: "navigation"
  }, (0, _element.createElement)("div", {
    className: "alignleft"
  }, (0, _element.createElement)("a", {
    href: "#top"
  }, "\xAB ", (0, _i18n.__)('Older Comments'))), (0, _element.createElement)("div", {
    className: "alignright"
  }, (0, _element.createElement)("a", {
    href: "#top"
  }, (0, _i18n.__)('Newer Comments'), " \xBB"))), (0, _element.createElement)(_form.default, {
    postId: postId,
    postType: postType
  }));
}
