// @wordpress/block-editor/build/hooks/index.native.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "getBorderClassesAndStyles", {
  enumerable: true,
  get: function () {
    return _useBorderProps.getBorderClassesAndStyles;
  }
});
Object.defineProperty(exports, "getColorClassesAndStyles", {
  enumerable: true,
  get: function () {
    return _useColorProps.getColorClassesAndStyles;
  }
});
Object.defineProperty(exports, "getSpacingClassesAndStyles", {
  enumerable: true,
  get: function () {
    return _useSpacingProps.getSpacingClassesAndStyles;
  }
});
Object.defineProperty(exports, "useBorderProps", {
  enumerable: true,
  get: function () {
    return _useBorderProps.useBorderProps;
  }
});
Object.defineProperty(exports, "useCachedTruthy", {
  enumerable: true,
  get: function () {
    return _useCachedTruthy.useCachedTruthy;
  }
});
Object.defineProperty(exports, "useColorProps", {
  enumerable: true,
  get: function () {
    return _useColorProps.useColorProps;
  }
});

require("./compat");

require("./align");

require("./anchor");

require("./custom-class-name");

require("./generated-class-name");

require("./style");

require("./color");

require("./font-size");

require("./layout");

var _useBorderProps = require("./use-border-props");

var _useColorProps = require("./use-color-props");

var _useSpacingProps = require("./use-spacing-props");

var _useCachedTruthy = require("./use-cached-truthy");
