// @wordpress/block-editor/build/hooks/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "getBorderClassesAndStyles", {
  enumerable: true,
  get: function () {
    return _useBorderProps.getBorderClassesAndStyles;
  }
});
Object.defineProperty(exports, "getColorClassesAndStyles", {
  enumerable: true,
  get: function () {
    return _useColorProps.getColorClassesAndStyles;
  }
});
Object.defineProperty(exports, "getGapCSSValue", {
  enumerable: true,
  get: function () {
    return _gap.getGapCSSValue;
  }
});
Object.defineProperty(exports, "getSpacingClassesAndStyles", {
  enumerable: true,
  get: function () {
    return _useSpacingProps.getSpacingClassesAndStyles;
  }
});
Object.defineProperty(exports, "getTypographyClassesAndStyles", {
  enumerable: true,
  get: function () {
    return _useTypographyProps.getTypographyClassesAndStyles;
  }
});
Object.defineProperty(exports, "useBorderProps", {
  enumerable: true,
  get: function () {
    return _useBorderProps.useBorderProps;
  }
});
Object.defineProperty(exports, "useCachedTruthy", {
  enumerable: true,
  get: function () {
    return _useCachedTruthy.useCachedTruthy;
  }
});
Object.defineProperty(exports, "useColorProps", {
  enumerable: true,
  get: function () {
    return _useColorProps.useColorProps;
  }
});
Object.defineProperty(exports, "useCustomSides", {
  enumerable: true,
  get: function () {
    return _dimensions.useCustomSides;
  }
});
Object.defineProperty(exports, "useLayoutClasses", {
  enumerable: true,
  get: function () {
    return _layout.useLayoutClasses;
  }
});
Object.defineProperty(exports, "useLayoutStyles", {
  enumerable: true,
  get: function () {
    return _layout.useLayoutStyles;
  }
});

require("./compat");

require("./align");

require("./lock");

require("./anchor");

require("./aria-label");

require("./custom-class-name");

require("./generated-class-name");

require("./style");

require("./settings");

require("./color");

require("./duotone");

require("./font-family");

require("./font-size");

require("./border");

require("./position");

var _layout = require("./layout");

require("./content-lock-ui");

require("./metadata");

require("./metadata-name");

var _dimensions = require("./dimensions");

var _useBorderProps = require("./use-border-props");

var _useColorProps = require("./use-color-props");

var _useSpacingProps = require("./use-spacing-props");

var _useTypographyProps = require("./use-typography-props");

var _gap = require("./gap");

var _useCachedTruthy = require("./use-cached-truthy");
