// @wordpress/block-editor/build/hooks/padding.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PaddingVisualizer = PaddingVisualizer;

var _element = require("@wordpress/element");

var _isShallowEqual = _interopRequireDefault(require("@wordpress/is-shallow-equal"));

var _blockPopover = _interopRequireDefault(require("../components/block-popover"));

var _utils = require("../components/spacing-sizes-control/utils");

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
function PaddingVisualizer(_ref) {
  var _attributes$style, _attributes$style$spa;

  let {
    clientId,
    attributes,
    forceShow
  } = _ref;
  const padding = attributes === null || attributes === void 0 ? void 0 : (_attributes$style = attributes.style) === null || _attributes$style === void 0 ? void 0 : (_attributes$style$spa = _attributes$style.spacing) === null || _attributes$style$spa === void 0 ? void 0 : _attributes$style$spa.padding;
  const style = (0, _element.useMemo)(() => {
    return {
      borderTopWidth: padding !== null && padding !== void 0 && padding.top ? (0, _utils.getSpacingPresetCssVar)(padding === null || padding === void 0 ? void 0 : padding.top) : 0,
      borderRightWidth: padding !== null && padding !== void 0 && padding.right ? (0, _utils.getSpacingPresetCssVar)(padding === null || padding === void 0 ? void 0 : padding.right) : 0,
      borderBottomWidth: padding !== null && padding !== void 0 && padding.bottom ? (0, _utils.getSpacingPresetCssVar)(padding === null || padding === void 0 ? void 0 : padding.bottom) : 0,
      borderLeftWidth: padding !== null && padding !== void 0 && padding.left ? (0, _utils.getSpacingPresetCssVar)(padding === null || padding === void 0 ? void 0 : padding.left) : 0
    };
  }, [padding]);
  const [isActive, setIsActive] = (0, _element.useState)(false);
  const valueRef = (0, _element.useRef)(padding);
  const timeoutRef = (0, _element.useRef)();

  const clearTimer = () => {
    if (timeoutRef.current) {
      window.clearTimeout(timeoutRef.current);
    }
  };

  (0, _element.useEffect)(() => {
    if (!(0, _isShallowEqual.default)(padding, valueRef.current) && !forceShow) {
      setIsActive(true);
      valueRef.current = padding;
      timeoutRef.current = setTimeout(() => {
        setIsActive(false);
      }, 400);
    }

    return () => {
      setIsActive(false);
      clearTimer();
    };
  }, [padding, forceShow]);

  if (!isActive && !forceShow) {
    return null;
  }

  return (0, _element.createElement)(_blockPopover.default, {
    clientId: clientId,
    __unstableCoverTarget: true,
    __unstableRefreshSize: padding,
    __unstablePopoverSlot: "block-toolbar",
    shift: false
  }, (0, _element.createElement)("div", {
    className: "block-editor__padding-visualizer",
    style: style
  }));
}
