// @wordpress/block-editor/build-module/components/block-draggable/draggable-chip.native.js
import { createElement } from "@wordpress/element";

/**
 * External dependencies
 */
import { View } from 'react-native';
/**
 * WordPress dependencies
 */

import { dragHandle } from '@wordpress/icons';
import { usePreferredColorSchemeStyle } from '@wordpress/compose';
/**
 * Internal dependencies
 */

import BlockIcon from '../block-icon';
import styles from './style.scss';
const shadowStyle = {
  shadowColor: '#000',
  shadowOffset: {
    width: 0,
    height: 2
  },
  shadowOpacity: 0.25,
  shadowRadius: 3.84,
  elevation: 5
};
/**
 * Block draggable chip component
 *
 * @param {Object} props        Component props.
 * @param {Object} [props.icon] Block icon.
 * @return {JSX.Element} Chip component.
 */

export default function BlockDraggableChip(_ref) {
  let {
    icon
  } = _ref;
  const containerStyle = usePreferredColorSchemeStyle(styles['draggable-chip__container'], styles['draggable-chip__container--dark']);
  return createElement(View, {
    style: [containerStyle, shadowStyle],
    testID: "draggable-chip"
  }, createElement(BlockIcon, {
    icon: dragHandle
  }), icon && createElement(BlockIcon, {
    icon: icon
  }));
}
