// @wordpress/block-editor/build/components/inspector-controls-tabs/settings-tab-hint.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = InspectorControlsTabsHint;

var _element = require("@wordpress/element");

var _components = require("@wordpress/components");

var _data = require("@wordpress/data");

var _dom = require("@wordpress/dom");

var _i18n = require("@wordpress/i18n");

var _icons = require("@wordpress/icons");

var _preferences = require("@wordpress/preferences");

/**
 * WordPress dependencies
 */
const PREFERENCE_NAME = 'isInspectorControlsTabsHintVisible';

function InspectorControlsTabsHint() {
  const isInspectorControlsTabsHintVisible = (0, _data.useSelect)(select => {
    var _select$get;

    return (_select$get = select(_preferences.store).get('core', PREFERENCE_NAME)) !== null && _select$get !== void 0 ? _select$get : true;
  }, []);
  const ref = (0, _element.useRef)();
  const {
    set: setPreference
  } = (0, _data.useDispatch)(_preferences.store);

  if (!isInspectorControlsTabsHintVisible) {
    return null;
  }

  return (0, _element.createElement)("div", {
    ref: ref,
    className: "block-editor-inspector-controls-tabs__hint"
  }, (0, _element.createElement)("div", {
    className: "block-editor-inspector-controls-tabs__hint-content"
  }, (0, _i18n.__)("Looking for other block settings? They've moved to the styles tab.")), (0, _element.createElement)(_components.Button, {
    className: "block-editor-inspector-controls-tabs__hint-dismiss",
    icon: _icons.close,
    iconSize: "16",
    label: (0, _i18n.__)('Dismiss hint'),
    onClick: () => {
      // Retain focus when dismissing the element.
      const previousElement = _dom.focus.tabbable.findPrevious(ref.current);

      previousElement === null || previousElement === void 0 ? void 0 : previousElement.focus();
      setPreference('core', PREFERENCE_NAME, false);
    },
    showTooltip: false
  }));
}
