// @wordpress/block-library/build/site-title/edit/level-toolbar.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = LevelControl;

var _element = require("@wordpress/element");

var _components = require("@wordpress/components");

var _i18n = require("@wordpress/i18n");

var _levelIcon = _interopRequireDefault(require("./level-icon"));

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
function LevelControl(_ref) {
  let {
    level,
    onChange
  } = _ref;
  const allControls = [1, 2, 3, 4, 5, 6, 0].map(currentLevel => {
    const isActive = currentLevel === level;
    return {
      icon: (0, _element.createElement)(_levelIcon.default, {
        level: currentLevel,
        isPressed: isActive
      }),
      title: currentLevel === 0 ? (0, _i18n.__)('Paragraph') : // translators: %s: heading level e.g: "1", "2", "3"
      (0, _i18n.sprintf)((0, _i18n.__)('Heading %d'), currentLevel),
      isActive,
      onClick: () => onChange(currentLevel),
      role: 'menuitemradio'
    };
  });
  return (0, _element.createElement)(_components.ToolbarDropdownMenu, {
    label: (0, _i18n.__)('Change heading level'),
    icon: (0, _element.createElement)(_levelIcon.default, {
      level: level
    }),
    controls: allControls
  });
}
