// @wordpress/block-editor/build-module/components/block-styles/preview-panel.js
import { createElement } from "@wordpress/element";

/**
 * WordPress dependencies
 */
import { getBlockType } from '@wordpress/blocks';
import { useMemo } from '@wordpress/element';
/**
 * Internal dependencies
 */

import InserterPreviewPanel from '../inserter/preview-panel';
import { replaceActiveStyle } from './utils';
export default function BlockStylesPreviewPanel(_ref) {
  var _getBlockType;

  let {
    genericPreviewBlock,
    style,
    className,
    activeStyle
  } = _ref;
  const example = (_getBlockType = getBlockType(genericPreviewBlock.name)) === null || _getBlockType === void 0 ? void 0 : _getBlockType.example;
  const styleClassName = replaceActiveStyle(className, activeStyle, style);
  const previewBlocks = useMemo(() => {
    return { ...genericPreviewBlock,
      title: style.label || style.name,
      description: style.description,
      initialAttributes: { ...genericPreviewBlock.attributes,
        className: styleClassName + ' block-editor-block-styles__block-preview-container'
      },
      example
    };
  }, [genericPreviewBlock, styleClassName]);
  return createElement(InserterPreviewPanel, {
    item: previewBlocks,
    isStylePreview: true
  });
}
