// @wordpress/block-editor/build/components/inspector-popover-header/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = InspectorPopoverHeader;

var _element = require("@wordpress/element");

var _components = require("@wordpress/components");

var _icons = require("@wordpress/icons");

var _i18n = require("@wordpress/i18n");

/**
 * WordPress dependencies
 */
function InspectorPopoverHeader(_ref) {
  let {
    title,
    help,
    actions = [],
    onClose
  } = _ref;
  return (0, _element.createElement)(_components.__experimentalVStack, {
    className: "block-editor-inspector-popover-header",
    spacing: 4
  }, (0, _element.createElement)(_components.__experimentalHStack, {
    alignment: "center"
  }, (0, _element.createElement)(_components.__experimentalHeading, {
    className: "block-editor-inspector-popover-header__heading",
    level: 2,
    size: 13
  }, title), (0, _element.createElement)(_components.__experimentalSpacer, null), actions.map(_ref2 => {
    let {
      label,
      icon,
      onClick
    } = _ref2;
    return (0, _element.createElement)(_components.Button, {
      key: label,
      className: "block-editor-inspector-popover-header__action",
      label: label,
      icon: icon,
      variant: !icon && 'tertiary',
      onClick: onClick
    }, !icon && label);
  }), onClose && (0, _element.createElement)(_components.Button, {
    className: "block-editor-inspector-popover-header__action",
    label: (0, _i18n.__)('Close'),
    icon: _icons.closeSmall,
    onClick: onClose
  })), help && (0, _element.createElement)(_components.__experimentalText, null, help));
}
