// @wordpress/block-library/build/comments-pagination-numbers/edit.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = CommentsPaginationNumbersEdit;

var _element = require("@wordpress/element");

var _blockEditor = require("@wordpress/block-editor");

/**
 * WordPress dependencies
 */
const PaginationItem = _ref => {
  let {
    content,
    tag: Tag = 'a',
    extraClass = ''
  } = _ref;
  return Tag === 'a' ? (0, _element.createElement)(Tag, {
    className: `page-numbers ${extraClass}`,
    href: "#comments-pagination-numbers-pseudo-link",
    onClick: event => event.preventDefault()
  }, content) : (0, _element.createElement)(Tag, {
    className: `page-numbers ${extraClass}`
  }, content);
};

function CommentsPaginationNumbersEdit() {
  return (0, _element.createElement)("div", (0, _blockEditor.useBlockProps)(), (0, _element.createElement)(PaginationItem, {
    content: "1"
  }), (0, _element.createElement)(PaginationItem, {
    content: "2"
  }), (0, _element.createElement)(PaginationItem, {
    content: "3",
    tag: "span",
    extraClass: "current"
  }), (0, _element.createElement)(PaginationItem, {
    content: "4"
  }), (0, _element.createElement)(PaginationItem, {
    content: "5"
  }), (0, _element.createElement)(PaginationItem, {
    content: "...",
    tag: "span",
    extraClass: "dots"
  }), (0, _element.createElement)(PaginationItem, {
    content: "8"
  }));
}
