// @wordpress/block-library/build/read-more/edit.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ReadMore;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _blockEditor = require("@wordpress/block-editor");

var _components = require("@wordpress/components");

var _blocks = require("@wordpress/blocks");

var _i18n = require("@wordpress/i18n");

/**
 * WordPress dependencies
 */
function ReadMore(_ref) {
  let {
    attributes: {
      content,
      linkTarget
    },
    setAttributes,
    insertBlocksAfter
  } = _ref;
  const blockProps = (0, _blockEditor.useBlockProps)();
  return (0, _element.createElement)(_element.Fragment, null, (0, _element.createElement)(_blockEditor.InspectorControls, null, (0, _element.createElement)(_components.PanelBody, {
    title: (0, _i18n.__)('Link settings')
  }, (0, _element.createElement)(_components.ToggleControl, {
    __nextHasNoMarginBottom: true,
    label: (0, _i18n.__)('Open in new tab'),
    onChange: value => setAttributes({
      linkTarget: value ? '_blank' : '_self'
    }),
    checked: linkTarget === '_blank'
  }))), (0, _element.createElement)(_blockEditor.RichText, (0, _extends2.default)({
    tagName: "a",
    "aria-label": (0, _i18n.__)('“Read more” link text'),
    placeholder: (0, _i18n.__)('Read more'),
    value: content,
    onChange: newValue => setAttributes({
      content: newValue
    }),
    __unstableOnSplitAtEnd: () => insertBlocksAfter((0, _blocks.createBlock)((0, _blocks.getDefaultBlockName)())),
    withoutInteractiveFormatting: true
  }, blockProps)));
}
