// @wordpress/block-editor/build-module/hooks/margin.js
import { createElement } from "@wordpress/element";

/**
 * WordPress dependencies
 */
import { useMemo, useRef, useState, useEffect } from '@wordpress/element';
import isShallowEqual from '@wordpress/is-shallow-equal';
/**
 * Internal dependencies
 */

import BlockPopover from '../components/block-popover';
import { getSpacingPresetCssVar } from '../components/spacing-sizes-control/utils';
export function MarginVisualizer(_ref) {
  var _attributes$style, _attributes$style$spa;

  let {
    clientId,
    attributes,
    forceShow
  } = _ref;
  const margin = attributes === null || attributes === void 0 ? void 0 : (_attributes$style = attributes.style) === null || _attributes$style === void 0 ? void 0 : (_attributes$style$spa = _attributes$style.spacing) === null || _attributes$style$spa === void 0 ? void 0 : _attributes$style$spa.margin;
  const style = useMemo(() => {
    const marginTop = margin !== null && margin !== void 0 && margin.top ? getSpacingPresetCssVar(margin === null || margin === void 0 ? void 0 : margin.top) : 0;
    const marginRight = margin !== null && margin !== void 0 && margin.right ? getSpacingPresetCssVar(margin === null || margin === void 0 ? void 0 : margin.right) : 0;
    const marginBottom = margin !== null && margin !== void 0 && margin.bottom ? getSpacingPresetCssVar(margin === null || margin === void 0 ? void 0 : margin.bottom) : 0;
    const marginLeft = margin !== null && margin !== void 0 && margin.left ? getSpacingPresetCssVar(margin === null || margin === void 0 ? void 0 : margin.left) : 0;
    return {
      borderTopWidth: marginTop,
      borderRightWidth: marginRight,
      borderBottomWidth: marginBottom,
      borderLeftWidth: marginLeft,
      top: marginTop ? `calc(${marginTop} * -1)` : 0,
      right: marginRight ? `calc(${marginRight} * -1)` : 0,
      bottom: marginBottom ? `calc(${marginBottom} * -1)` : 0,
      left: marginLeft ? `calc(${marginLeft} * -1)` : 0
    };
  }, [margin]);
  const [isActive, setIsActive] = useState(false);
  const valueRef = useRef(margin);
  const timeoutRef = useRef();

  const clearTimer = () => {
    if (timeoutRef.current) {
      window.clearTimeout(timeoutRef.current);
    }
  };

  useEffect(() => {
    if (!isShallowEqual(margin, valueRef.current) && !forceShow) {
      setIsActive(true);
      valueRef.current = margin;
      timeoutRef.current = setTimeout(() => {
        setIsActive(false);
      }, 400);
    }

    return () => {
      setIsActive(false);
      clearTimer();
    };
  }, [margin, forceShow]);

  if (!isActive && !forceShow) {
    return null;
  }

  return createElement(BlockPopover, {
    clientId: clientId,
    __unstableCoverTarget: true,
    __unstableRefreshSize: margin,
    __unstablePopoverSlot: "block-toolbar",
    shift: false
  }, createElement("div", {
    className: "block-editor__padding-visualizer",
    style: style
  }));
}
