// @wordpress/block-editor/build/components/block-alignment-control/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.POPOVER_PROPS = exports.DEFAULT_CONTROL = exports.BLOCK_ALIGNMENTS_CONTROLS = void 0;

var _i18n = require("@wordpress/i18n");

var _icons = require("@wordpress/icons");

/**
 * WordPress dependencies
 */
const BLOCK_ALIGNMENTS_CONTROLS = {
  none: {
    icon: _icons.alignNone,
    title: (0, _i18n._x)('None', 'Alignment option')
  },
  left: {
    icon: _icons.positionLeft,
    title: (0, _i18n.__)('Align left')
  },
  center: {
    icon: _icons.positionCenter,
    title: (0, _i18n.__)('Align center')
  },
  right: {
    icon: _icons.positionRight,
    title: (0, _i18n.__)('Align right')
  },
  wide: {
    icon: _icons.stretchWide,
    title: (0, _i18n.__)('Wide width')
  },
  full: {
    icon: _icons.stretchFullWidth,
    title: (0, _i18n.__)('Full width')
  }
};
exports.BLOCK_ALIGNMENTS_CONTROLS = BLOCK_ALIGNMENTS_CONTROLS;
const DEFAULT_CONTROL = 'none';
exports.DEFAULT_CONTROL = DEFAULT_CONTROL;
const POPOVER_PROPS = {
  variant: 'toolbar'
};
exports.POPOVER_PROPS = POPOVER_PROPS;
