// @wordpress/block-library/build/query-pagination-next/edit.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = QueryPaginationNextEdit;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _i18n = require("@wordpress/i18n");

var _blockEditor = require("@wordpress/block-editor");

/**
 * WordPress dependencies
 */
const arrowMap = {
  none: '',
  arrow: '→',
  chevron: '»'
};

function QueryPaginationNextEdit(_ref) {
  let {
    attributes: {
      label
    },
    setAttributes,
    context: {
      paginationArrow
    }
  } = _ref;
  const displayArrow = arrowMap[paginationArrow];
  return (0, _element.createElement)("a", (0, _extends2.default)({
    href: "#pagination-next-pseudo-link",
    onClick: event => event.preventDefault()
  }, (0, _blockEditor.useBlockProps)()), (0, _element.createElement)(_blockEditor.PlainText, {
    __experimentalVersion: 2,
    tagName: "span",
    "aria-label": (0, _i18n.__)('Next page link'),
    placeholder: (0, _i18n.__)('Next Page'),
    value: label,
    onChange: newLabel => setAttributes({
      label: newLabel
    })
  }), displayArrow && (0, _element.createElement)("span", {
    className: `wp-block-query-pagination-next-arrow is-arrow-${paginationArrow}`,
    "aria-hidden": true
  }, displayArrow));
}
