// @wordpress/block-editor/build/components/spacing-sizes-control/all-input-control.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = AllInputControl;

var _element = require("@wordpress/element");

var _components = require("@wordpress/components");

var _spacingInputControl = _interopRequireDefault(require("./spacing-input-control"));

var _utils = require("./utils");

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
function AllInputControl(_ref) {
  let {
    onChange,
    values,
    sides,
    spacingSizes,
    type,
    minimumCustomValue,
    onMouseOver,
    onMouseOut
  } = _ref;
  const allValue = (0, _utils.getAllRawValue)(values);
  const hasValues = (0, _utils.isValuesDefined)(values);
  const isMixed = hasValues && (0, _utils.isValuesMixed)(values, sides);

  const handleOnChange = next => {
    const nextValues = (0, _components.__experimentalApplyValueToSides)(values, next, sides);
    onChange(nextValues);
  };

  return (0, _element.createElement)(_spacingInputControl.default, {
    value: allValue,
    onChange: handleOnChange,
    side: 'all',
    spacingSizes: spacingSizes,
    isMixed: isMixed,
    type: type,
    minimumCustomValue: minimumCustomValue,
    onMouseOver: onMouseOver,
    onMouseOut: onMouseOut
  });
}
