// @wordpress/block-editor/build/components/inner-blocks/button-block-appender.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.ButtonBlockAppender = void 0;

var _element = require("@wordpress/element");

var _classnames = _interopRequireDefault(require("classnames"));

var _buttonBlockAppender = _interopRequireDefault(require("../button-block-appender"));

var _withClientId = _interopRequireDefault(require("./with-client-id"));

/**
 * External dependencies
 */

/**
 * Internal dependencies
 */
const ButtonBlockAppender = _ref => {
  let {
    clientId,
    showSeparator,
    isFloating,
    onAddBlock,
    isToggle
  } = _ref;
  return (0, _element.createElement)(_buttonBlockAppender.default, {
    className: (0, _classnames.default)({
      'block-list-appender__toggle': isToggle
    }),
    rootClientId: clientId,
    showSeparator: showSeparator,
    isFloating: isFloating,
    onAddBlock: onAddBlock
  });
};

exports.ButtonBlockAppender = ButtonBlockAppender;

var _default = (0, _withClientId.default)(ButtonBlockAppender);

exports.default = _default;
